Release `3.3.0` with backwards compatibility fixes (#1284)
paritytech/ink: RELEASES.md

# [Unreleased]
- Backward-incompatible piece of [#1233](https://github.com/paritytech/ink/pull/1233): removal of `eth_compatibility crate`
- Backward-incompatible piece of [#1224](https://github.com/paritytech/ink/pull/1224): dependency on `[seal1] seal_set_storage`
- Backward-incompatible "Optimise deny_payment. Use eerywhere semantic of deny. ([#1267](https://github.com/paritytech/ink/pull/1267))"

>   (see @HCastano SE [answer](https://substrate.stackexchange.com/a/3000/472) in this regard)

Reviewer: Not sure if we need to mention these changes here

Author: if we go with separate 3.x.x branch, I agree. This is done not to forget mentioning them in the 4.x release

# Version 3.3.0

> This release is to make *ink!* 3.x.x backward compatible with *substrate-contracts-node* version [0.13.0](https://github.com/paritytech/substrate-contracts-node/releases/tag/v0.13.0) again. 

Reviewer: This is just one of the reasons, we should also mention that we didn't follow SemVer

Reviewer: We're trying to alleviate two things here, incompatibility between 3.2 and earlier 3.x releases, as well as incompatibility with the 0.13 node

> This release is to make *ink!* 3.x.x backward compatible with *substrate-contracts-node* version [0.13.0](https://github.com/paritytech/substrate-contracts-node/releases/tag/v0.13.0) again. 

Reviewer 2: ```suggestion
This release is to make ink! 3.x.x backwards compatible with `substrate-contracts-node` version [0.13.0](https://github.com/paritytech/substrate-contracts-node/releases/tag/v0.13.0) again. 
```

## Compatibility

> This version should work fine with *substrate-contracts-node* versions from [0.13.0](https://github.com/paritytech/substrate-contracts-node/releases/tag/v0.13.0) up to  [0.16.0](https://github.com/paritytech/substrate-contracts-node/releases/tag/v0.16.0)

Reviewer: ```suggestion
This version should work fine with *substrate-contracts-node* versions from [0.13.0](https://github.com/paritytech/substrate-contracts-node/releases/tag/v0.13.0) up to [0.16.0](https://github.com/paritytech/substrate-contracts-node/releases/tag/v0.16.0)
```

## Changed
*Context: user-reported issues on our SE ([first](https://substrate.stackexchange.com/questions/2721/cargo-contract-3-0-1) and [second](https://substrate.stackexchange.com/questions/2870/cargo-contract-throws-error-about-supplied-arguments-in-inkconstructor-f)) unveiled backward incompatibility introduced in 3.1.0 release.*

The following has been done to restore backward compatibility:

> - Reverted backward-incompatible piece of [#1233](https://github.com/paritytech/ink/pull/1233): removal of `eth_compatibility crate`

Reviewer: ```suggestion
- Reverted backward-incompatible piece of [#1233](https://github.com/paritytech/ink/pull/1233). This removed the `eth_compatibility crate`.
```

> - Reverted backward-incompatible piece of [#1224](https://github.com/paritytech/ink/pull/1224): dependency on `[seal1] seal_set_storage`

Reviewer: ```suggestion
- Reverted backward-incompatible piece of [#1224](https://github.com/paritytech/ink/pull/1224). An old method depended on a new SEAL API (`[seal1] seal_set_storage`).
```

> - Reverted "Optimise deny_payment. Use eerywhere semantic of deny. ([#1267](https://github.com/paritytech/ink/pull/1267))"  

Reviewer: ```suggestion
- Reverted "Optimise deny_payment. Use everywhere semantic of deny. ([#1267](https://github.com/paritytech/ink/pull/1267))"  
```

Reviewer: Sure, but it's a typo and we shouldn't propagate that everywhere 😄

>   (this one is to restore compatibiliy between minor versions of ink! crates; see @HCastano SE [answer](https://substrate.stackexchange.com/a/3000/472) in this regard)

Reviewer: ```suggestion
  This one is to restore compatibility between minor versions of ink! crates; see @HCastano's SE [answer](https://substrate.stackexchange.com/a/3000/472) in this regard.
```
